p2panda-net/src/sync/manager.rs

```diff
@@ -332,26 +332,23 @@ where
             session.remove(&sync_attempt.peer);
         }
 
-        if let Some(err) = err.downcast_ref() {
-            match err {
-                // If the sync attempt failed because of a connection error we want to retry up to
-                // `max_retry_attempts`. If error occurs after this we simply stop trying without
-                // informing the engine as it never knew the attempts were occurring.
-                SyncAttemptError::Connection => {
-                    warn!("sync attempt failed due to connection error");
-                    if sync_attempt.attempts <= self.config.max_retry_attempts {
-                        self.reschedule_attempt(sync_attempt).await?;
-                        return Ok(());
-                    }
-                }
-                SyncAttemptError::Sync(_) => {
-                    self.engine_actor_tx
-                        .send(ToEngineActor::SyncFailed {
-                            topic: Some(sync_attempt.topic),
-                            peer: sync_attempt.peer,
-                        })
-                        .await?;
-                }
+        if let Some(err) = err.downcast_ref::<SyncAttemptError>() {
+            // If the sync attempt failed for any reason we want to retry up to
+            // `max_retry_attempts`.
+            warn!("sync attempt failed: {err}");
+
+            // We need to inform the engine of the failed attempt so that the gossip buffer counter
+            // can be decremented.
+            self.engine_actor_tx
```

Comment:
I invite careful consideration of this behaviour.

The context here is a failed sync session from the perspective of the initiator (we are in the sync manager where connections and sessions are initiated). Since the topic is known by the initiator, a `SyncStart` event will have been sent to the engine (this happens in `p2panda-net/src/sync/initiate.rs`). This results in a locking of the gossip buffer (thereby incrementing the counter). As a result, we need to send `SyncFailed` here to ensure the gossip buffer counter is correctly decremented.